### ISSUE TITLE
JSON parsing/serialization for Avro types

### ISSUE THREAD

**Issue by TikhonJelvis** [GaloisInc/avro#15]
Is there a convenient way to read/write types with `FromAvro`/`ToAvro` instances as JSON rather than Avro's binary format, using the corresponding Avro schema?

This would be particularly nice with the types generated by `deriveAvro` as it would let me easily support both formats.

If there's no easy way to do this as is, what would adding that to the library look like? Would a good option be to move from `Avro.Value` to `Aeson.Value`?

I'm up for implementing a PR if it isn't too tricky :).

**Comment by AlexeyRaga** [comment]
You can decode any Avro value into `Aeson.Value` by just using `decodeAvro`, but you still need schema because without schema Avro decoder doesn't know what to do.
`ToJSON` instance for `Data.Avro.Types.Value` already exists.

```
decodeAvro mySchema myPayload :: Either String Aeson.Value
```

I don't know about decoding **from** JSON though. I think it should be possible: knowing the schema you could traverse `Aeson.Value` and create `Data.Avro.Types.Value`. So it should be possible to write a generic function like `fromJSON :: HasAvroSchema a => Aeson.JSON -> a`.
Maybe even `FromJSON` instance for `Data.Avro.Types.Value`.

However, for both operations you would still need schema. Without Schema it isn't possible to determine types of fields.
For example, when JSON says:
```
{
    "severity" : 12,
    "mitigationType": null 
}
```

then only schema can tell us that `severity` should be of type `Double` and not `Int32`, that `mitigationType` is of type `Maybe Mitigation`, etc.

So once you say that the schema is required (or the result type is known as the schema can be obtained from it) then it shouldn't be too tricky.

**Comment by AlexeyRaga** [comment]
Yeah, I think that `HasAvroSchema a => Aeson.JSON -> Either String a` should be possible.
Try making it happen, maybe looking at . `ToJSON` as an example: https://github.com/GaloisInc/avro/blob/master/src/Data/Avro/Schema.hs#L307

**Comment by TikhonJelvis** [comment]
@AlexeyRaga Oh hey, there's a [`parseAvroJSON`](https://github.com/GaloisInc/avro/blob/master/src/Data/Avro/Schema.hs#L366) function in `Data.Avro.Schema`, but not exported. 

Doesn't this do almost exactly what I want? It seems like I can just compose it with `fromAvro`:

```
fromJSON :: HasAvroSchema a => Aeson.JSON -> Result a
fromJSON json = parseAvroJSON env schema json >>= fromAvro
  where {- ... -}
```

**Comment by TikhonJelvis** [comment]
Of course, I found this function after I was most of the way through writing my own version of `parseAvroJSON` :).

**Comment by AlexeyRaga** [comment]
@TikhonJelvis I wouldn't know for sure because I haven't used it, but it does look what you need! Good to know!

Perhaps some unit/property testing would be nice to have to make sure that it works properly :)

**Comment by TikhonJelvis** [comment]
I'll give it a try.

I did figure out one awkward thing though: `parseAvroJSON` depends on `Type` from `Data.Avro.Schema` and the `FromJSON` instance of `Type` depends on `parseAvroJSON`. This means we can't move `parseAvroJSON` into a separate module without incurring a circular dependency :/.

One option would be moving `Type` into `Data.Avro.Types` and reexporting it from `Data.Avro.Schema`. How does that sound?

**Comment by AlexeyRaga** [comment]
Yes,  it can be cleaned up a bit in my opinion too.
I am not against moving `Type` and its synonym `Schema` to `Data.Avro.Types`. I think it is quite logical.

I would even turn them around and made `Schema` to be a real data type name, and had type `Type` a synonym so we could eventually deprecate `Type` and just use `Schema` everywhere :)   
But these are just my thoughts.

**Comment by TikhonJelvis** [comment]
> I would even turn them around and made `Schema` to be a real data type name

I was going to suggest exactly that! `Type` is too generic...

**Comment by TikhonJelvis** [comment]
Actually, never mind: moving `Type` to `Data.Avro.Types` wouldn't solve my problem—I'd still either have to put `parseAvroJSON` into `Data.Avro.Types` or have circular dependencies or have an orphan `FromJSON` instance. None of these are great choices.

This is one of those cases where the open-world assumption really bites us :/.

**Comment by TikhonJelvis** [comment]
Okay, I've written code for this, now I just need to write some tests :).

My change might actually not be backwards compatible with the library because the previous version was too permissive in parsing default values for unions. The spec only allows the default value to be of the *first* schema in a union, but the existing code tries *all* of the specs in the union:

> Default values for union fields correspond to the first schema in the union.

(See https://avro.apache.org/docs/1.8.2/spec.html#schema_record)

By the way, is this the right repo to open a PR?

**Comment by AlexeyRaga** [comment]
> By the way, is this the right repo to open a PR?

From my understanding, @TomMD is not actively interested in this lib at the moment, and is not actively reviewing/merging PRs.
I don't have any special rights on this repo so all the patches I and my colleagues make are going through this repo: https://github.com/haskell-works/hw-haskell-avro
We always send PRs to this repo, but we have no way to tell when and if they'll be accepted. We also publish to Hackage from `haskell-works`.

You can make your PR against https://github.com/haskell-works/hw-haskell-avro and then it will probably be merged and end up on Hackage a bit quicker.

**Comment by TomMD** [comment]
@AlexeyRaga's understanding is correct.  I am happy to give the commit bit on this repo to any interested contributors but do not currently have the time to manage the project.

**Comment by TikhonJelvis** [comment]
I don't have strong preferences between the repos, but it would be really helpful if one or the other was clearly marked as the "official" repo for the package. This means either:

  * move development to this repo and give a commit bit to @AlexeyRaga and whoever else is needed
  * keep active development on the haskell-works repo and have a link from this repo to that one

**Comment by TikhonJelvis** [comment]
It's much easier to contribute to a repo if all the issues and PRs are in one place.

Also, I noticed that the Hackage page links to *both* repos, which doesn't help with the confusion :).

**Comment by AlexeyRaga** [comment]
@TomMD If we either have commit rights on this repo or if we transfer this repo to `haskell-works` and kill that fork (obviously keeping the license, copyrights, etc.), both solutions work for me. 

If you don't think that you are going to be back to this project then maybe transfer works a bit better, but again, I am happy with either.

**Comment by TomMD** [comment]
I thought you did have commit rights... checks... hurm ok.  Well you now have an invite as admin on this repository.  Feel free to invite people as you see fit.  If I catch flak for giving away "admin" rights or there's a compelling reason then we can move the code, I'm not sensitive to the URL it's at.

**Comment by AlexeyRaga** [comment]
Thanks @TomMD! I'll do releases from this repo from now on.

**Comment by TikhonJelvis** [comment]
JSON parsing and serialization has been implemented in #21.
